flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/Schema.java

```diff
@@ -0,0 +1,856 @@
+         * functions (also defined in different catalogs) are supported.
+         *
+         * <p>Example: {@code .watermark("ts", $("ts).minus(lit(5).seconds())}
+         *
+         * @param columnName the column name used as a rowtime attribute
+         * @param watermarkExpression the expression used for watermark generation
+         */
+        public Builder watermark(String columnName, Expression watermarkExpression) {
+            Preconditions.checkNotNull(columnName, "Column name must not be null.");
+            Preconditions.checkNotNull(
+                    watermarkExpression, "Watermark expression must not be null.");
+            this.watermarkSpecs.add(new UnresolvedWatermarkSpec(columnName, watermarkExpression));
+            return this;
+        }
+
+        /**
+         * Declares that the given column should serve as an event-time (i.e. rowtime) attribute and
+         * specifies a corresponding watermark strategy as an expression.
+         *
+         * <p>See {@link #watermark(String, Expression)} for a detailed explanation.
+         *
+         * <p>This method uses a SQL expression that can be easily persisted in a durable catalog.
+         *
+         * <p>Example: {@code .watermark("ts", "ts - INTERVAL '5' SECOND")}
+         */
+        public Builder watermark(String columnName, String sqlExpression) {
+            return watermark(columnName, new SqlCallExpression(sqlExpression));
+        }
+
+        /**
+         * Declares a primary key constraint for a set of given columns. Primary key uniquely
+         * identify a row in a table. Neither of columns in a primary can be nullable. The primary
+         * key is informational only. It will not be enforced. It can be used for optimizations. It
+         * is the data owner's responsibility to ensure uniqueness of the data.
+         *
+         * <p>The primary key will be assigned a random name.
+         *
+         * @param columnNames columns that form a unique primary key
+         */
+        public Builder primaryKey(String... columnNames) {
+            Preconditions.checkNotNull(columnNames, "Primary key column names must not be null.");
+            return primaryKey(Arrays.asList(columnNames));
+        }
+
+        /**
+         * Declares a primary key constraint for a set of given columns. Primary key uniquely
+         * identify a row in a table. Neither of columns in a primary can be nullable. The primary
+         * key is informational only. It will not be enforced. It can be used for optimizations. It
+         * is the data owner's responsibility to ensure uniqueness of the data.
+         *
+         * <p>The primary key will be assigned a random name.
+         *
+         * @param columnNames columns that form a unique primary key
+         */
+        public Builder primaryKey(List<String> columnNames) {
+            return primaryKeyNamed(UUID.randomUUID().toString(), columnNames);
```

Comment:
We have an issue FLINK-16079 which wants to have a deterministic primary key name. What do you think about generate the primary key name by concat the field names, e.g. `PK_f0_f2`. This is also how other databases do.

Reply:
Sure, sounds good. I will implement that.